src/storage.rs

```diff
@@ -0,0 +1,189 @@
+use crate::app::{AppData, Channel, Message};
+use crate::util::StatefulList;
+
+use anyhow;
+use byteorder::{BigEndian, LittleEndian};
+use chrono::Utc;
+use sled;
+use unicode_width::UnicodeWidthStr;
+use zerocopy::{byteorder::U64, AsBytes, FromBytes, LayoutVerified, Unaligned, U16, U32};
+
+use std::fs::File;
+use std::path::Path;
+use std::str;
+
+pub trait Storage {
+    fn save(data: &AppData, path: impl AsRef<Path>) -> anyhow::Result<()>;
+
+    fn load(path: impl AsRef<Path>) -> anyhow::Result<AppData>;
+}
+
+pub struct Json;
+impl Storage for Json {
+    fn save(data: &AppData, path: impl AsRef<Path>) -> anyhow::Result<()> {
+        let f = File::create(path)?;
+        serde_json::to_writer(f, data)?;
+        Ok(())
+    }
+
+    fn load(path: impl AsRef<Path>) -> anyhow::Result<AppData> {
+        let f = File::open(path)?;
+        let mut data: AppData = serde_json::from_reader(f)?;
+        data.input_cursor = data.input.width();
+        Ok(data)
+    }
+}
+
+pub struct Db;
+
+impl Db {
+    fn load_channels(db: &sled::Db) -> anyhow::Result<Vec<Channel>> {
+        let channels = db.open_tree(b"channels")?;
+
+        let mut out = vec![];
+
+        for name_value_res in &channels {
+            let (_id, bytes) = name_value_res?;
+            let content = str::from_utf8(&bytes)?;
+            let channel = Channel {
+                id: "".to_owned(),
+                name: content.to_owned(),
+                is_group: false,
+                messages: Vec::new(),
+                unread_messages: 0,
+            };
+            out.push(channel);
+        }
+
+        Ok(out)
+    }
+
+    fn load_messages(db: &sled::Db) -> anyhow::Result<Vec<(U64<BigEndian>, Message)>> {
+        let messages = db.open_tree(b"messages")?;
+
+        let mut out = vec![];
+
+        for name_value_res in &messages {
+            let (_id, bytes) = name_value_res?;
+
+            let (channel_id, content): (_, LayoutVerified<U64<BigEndian>, &[u8]>) =
+                LayoutVerified::new_from_suffix(&*bytes).unwrap();
+            let content = str::from_utf8(&bytes)?;
+            let message = Message {
+                from: "not saved".to_owned(),
+                message: Some(content.to_owned()),
+                attachments: Vec::new(),
+                arrived_at: Utc::now(),
+            };
```

Comment:
I started following the example https://github.com/spacejam/sled/blob/master/examples/structured.rs. However, I'm not sure how to extract the channel Id followed by the string.